File: solutions/scala/luhn/1/src/main/scala/Luhn.scala
object Luhn {
  def valid(input: String): Boolean = {
    val cleanInput = input.replace(" ", "")
    if (isInvalidInput(cleanInput)) {
      false
    }
    else {
      val sum = cleanInput
        .reverse
        .split("")
        .zipWithIndex
        .map(double)
        .map(reduceOverage)
        .sum

      sum % 10 == 0
    }
  }

  private def isInvalidInput(cleanInput: String) =
    cleanInput.length <= 1 || !cleanInput.matches("\\d+")

  private def double(characterWithIndex: (String, Int)): Int = {
    val (character, index) = characterWithIndex
    val value = character.toInt
    if (this.isEven(index + 1)) {
      value * 2
    }
    else {
      value
    }
  }

  private def isEven(i: Int) =
    i % 2 == 0

  private def reduceOverage(i: Int): Int =
    if (i > 9) {
      i - 9
    }
    else {
      i
    }
}

File: solutions/elixir/luhn/1/lib/luhn.ex
defmodule Luhn do
  require Integer

  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?(number) do
    clean_input = String.replace(number, " ", "")

    valid_input?(clean_input) and
      rem(sum_of(clean_input), 10) === 0
  end

  defp valid_input?(clean_input),
    do:
      String.length(clean_input) > 1 and
        Regex.match?(~r/^\d+$/, clean_input)

  defp sum_of(clean_input),
    do:
      clean_input
      |> String.reverse()
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.map(&double/1)
      |> Enum.map(&reduce_overage/1)
      |> Enum.sum()

  defp double({character, index}),
    do:
      character
      |> String.to_integer()
      |> double(index)

  defp double(value, index)
       when Integer.is_odd(index),
       do: value * 2

  defp double(value, _index),
    do: value

  defp reduce_overage(integer)
       when integer > 9,
       do: integer - 9

  defp reduce_overage(integer),
    do: integer
end
